NWNXLib/Services/PerObjectStorage/PerObjectStorage.cpp

```diff
@@ -0,0 +1,224 @@
+{
+    if (!m_IntMap)
+        m_IntMap = std::make_unique<IntMap>();
+    return *m_IntMap;
+}
+PerObjectStorage::ObjectStorage::FloatMap& PerObjectStorage::ObjectStorage::GetFloatMap()
+{
+    if (!m_FloatMap)
+        m_FloatMap = std::make_unique<FloatMap>();
+    return *m_FloatMap;
+}
+PerObjectStorage::ObjectStorage::StringMap& PerObjectStorage::ObjectStorage::GetStringMap()
+{
+    if (!m_StringMap)
+        m_StringMap = std::make_unique<StringMap>();
+    return *m_StringMap;
+}
+PerObjectStorage::ObjectStorage::PointerMap& PerObjectStorage::ObjectStorage::GetPointerMap()
+{
+    if (!m_PointerMap)
+        m_PointerMap = std::make_unique<PointerMap>();
+    return *m_PointerMap;
+}
+
+PerObjectStorage::ObjectStorage::ObjectStorage(API::Types::ObjectID owner)
+{
+    m_oidOwner = owner;
+}
+PerObjectStorage::ObjectStorage::~ObjectStorage()
+{
+    if (m_PointerMap)
+    {
+        for (auto it: *m_PointerMap)
+        {
+            auto ptr = it.second.first;
+            auto cleanup = it.second.second;
+            cleanup(ptr);
+        }
+    }
+}
+
+
+
+PerObjectStorageProxy::PerObjectStorageProxy(PerObjectStorage& perObjectStorage, std::string pluginName)
+    : ServiceProxy<PerObjectStorage>(perObjectStorage)
+{
+    m_pluginName = pluginName;
+}
+PerObjectStorageProxy::~PerObjectStorageProxy()
+{
+    // TODO cleanup all storage from this plugin
+}
+
+void PerObjectStorageProxy::Set(API::Types::ObjectID object, std::string key, int value)
+{
+    m_proxyBase.Set(object, m_pluginName + "!" + key, value);
+}
+void PerObjectStorageProxy::Set(API::Types::ObjectID object, std::string key, float value)
+{
+    m_proxyBase.Set(object, m_pluginName + "!" + key, value);
+}
+void PerObjectStorageProxy::Set(API::Types::ObjectID object, std::string key, std::string value)
+{
+    m_proxyBase.Set(object, m_pluginName + "!" + key, value);
+}
+void PerObjectStorageProxy::Set(API::Types::ObjectID object, std::string key, void *value, PerObjectStorage::CleanupFunc cleanup)
+{
+    m_proxyBase.Set(object, m_pluginName + "!" + key, value, cleanup);
+}
+
+void PerObjectStorageProxy::Remove(API::Types::ObjectID object, std::string key)
+{
+    m_proxyBase.Remove(object, m_pluginName + "!" + key);
+}
+
+
+template <> Maybe<int> PerObjectStorage::Get<int>(API::Types::ObjectID object, std::string key)
+{
+    auto it = g_objectStorage.find(object);
+    if (it != g_objectStorage.end())
+    {
+        auto map = it->second->GetIntMap();
+        auto it2 = map.find(key);
+        if (it2 != map.end())
+            return Maybe<int>(std::move(it2->second));
+    }
+    return Maybe<int>();
+}
+template <> Maybe<float> PerObjectStorage::Get<float>(API::Types::ObjectID object, std::string key)
+{
+    auto it = g_objectStorage.find(object);
+    if (it != g_objectStorage.end())
+    {
+        auto map = it->second->GetFloatMap();
+        auto it2 = map.find(key);
+        if (it2 != map.end())
+            return Maybe<float>(std::move(it2->second));
```

Comment:
If you do this, when you call Get<>, it's going to steal the value from the map and move it into the Maybe. You don't want move here. Better to add an lvalue constructor to Maybe. I'll do it soon for you, then you can update this.

Reply:
_Really_? `Maybe<T>` has a `T m_data`, so I assume it would create a copy in the constructor when it does ` m_data(data)`

Reply:
Actually, you're right, but only because of a bug.

```cpp
template <typename T>
Maybe<T>::Maybe(T&& data)
    : m_hasValue(true), m_data(data)
{ }
```

It takes an rvalue reference. The rvalue reference decays into an lvalue reference, which is then copied with `m_data(data)`. What SHOULD happen is this: `m_data(std::forward<T>(data))`, which moves the value instead.

See https://github.com/nwnxee/unified/pull/77

Reply:
Updated with the new Maybe constructor..